### ISSUE TITLE
Changing Hosts is Destroying All Networked Objects

### ISSUE THREAD

**Issue by agincel** [Facepunch/sbox-issues#7226]
### Describe the bug

Starting very recently, any time the Host leaves in Marble Racer, the entire lobby state explodes. All player marbles, the game state controller, and anything else that's a Networked Object is destroyed, at which point the Lobby is irrecoverably broken.

What I BELIEVE is happening is that all Networked Objects, upon Host migration, are being treated as if they were owned by the previous Host, and their Orphaned Mode is treated as being "Destroy", regardless of actual setting.

I've tried several different publishes of Marble Racer from the past month, and all of them now exhibit this behavior when they hadn't previously, which leads me to believe that this is a s&box issue, and not some issue in my own configuration.

### To Reproduce

1. Have two players join Marble Racer
2. Have the host leave the game
3. Observe all networked objects destroy themselves, rendering the lobby unplayable.

### Expected behavior

Host migration should handle the transfer of networked objects as it did previously, obeying Orphaned Mode settings, and not destroying networked objects which should not be destroyed as part of the transfer.

### Media/Files

![Image](https://github.com/user-attachments/assets/db337a58-cc2b-4351-9ea9-97aee57ab261)

**Comment by aylaylay** [comment]
I have a feeling this might be related to json serialization issues. Have you tried going through your scene and making sure all objects have NetworkOrphaned set correctly? It could have been reset to default which is Destroy.

**Comment by agincel** [comment]
> I have a feeling this might be related to json serialization issues. Have you tried going through your scene and making sure all objects have NetworkOrphaned set correctly? It could have been reset to default which is Destroy.

Here are screenshots of the settings on all networked objects used in Marble Racer as of the most recent publish (the live version):

**Game State Controller**
![Image](https://github.com/user-attachments/assets/e8505623-9755-4954-83aa-05f03be3d6f8)

**Player Container**
![Image](https://github.com/user-attachments/assets/4a87d222-59e2-4cd2-be60-db53f32126fe)

...that's it. Player Container is a prefab spawned by Network Helper, which naturally sets the owner of each of these Player Containers to be the connection that spawned them. Yet, upon a host disconnecting, *all* marbles are destroyed. Plus the Game State Controller. It's really weird...

**Comment by kurozael** [comment]
Do you know roughly when this behaviour started?

**Comment by agincel** [comment]
I've added @kurozael and @aylaylay to the newly created Github Marble Racer repo. Let me know how else I can help - totally understood that something weird is going on. Enjoy your holiday!

**Comment by agincel** [comment]
> Do you know roughly when this behaviour started? Unfortunately I'm away for 5 days but will be able to investigate as soon as I return, as Layla said it could somehow be related to JSON serialisation issues but I'm sure this doesn't happen in the Networking test scene or the Walker game - not saying this isn't a s&box issue because it probably is, but I will need to narrow down why it's happening in your game. Are you able to give me access to the GitHub repo if I'm not already on it, I think you may have given me access before actually.

Also, first time I observed this behavior was on Friday 12/20, but I didn't play the past few days.

**Comment by aylaylay** [comment]
Is it possible to test this locally or do you need someone to join?

**Comment by agincel** [comment]
@aylaylay When I test locally, and I kill the hosting instance, I am seeing the behavior where the client's marble is erroneously deleted. That said, it's hard to test everything, since the host never truly migrates. That said, it seems like this issue is replicable with local testing.

TL;DR: in this video, the marble getting deleted on the client instance on the right should not be happening, and I cannot determine why it is.

https://streamable.com/cudcbs

**Comment by trundlr** [comment]
We are getting this in Micro Mayhem too

**Comment by MrBrax** [comment]
pretty sure host handover doesn't work properly when testing locally?

**Comment by agincel** [comment]
> pretty sure host handover doesn't work properly when testing locally?

You're absolutely right that it doesn't, but I'm still seeing this change in behavior. It used to be that I'd kill the host, and the orphaned clients would still have their marbles, but the networked logic would stop since the Host was gone. Now, instead, the client marbles are destroyed, which goes against what I have the settings set to.

**Comment by agincel** [comment]
I've got an even simpler repo, I'm pretty sure.

This project uses an unmodified Network Helper to spawn a unique Player for each new connection. I made that Player a rigidbody cube, and gave it a basic camera controller, just to make it easy to see what you control, though these are optional.

The client can leave and rejoin and things all work as expected, but upon the Host leaving, the Networked cubes which are owned by the clients are destroyed. This is incorrect behavior, as these objects were not owned by the Host.

https://streamable.com/atzvzo

[simplest_networking.zip](https://github.com/user-attachments/files/18237197/simplest_networking.zip)

**Comment by agincel** [comment]
[simplest_networking.zip](https://github.com/user-attachments/files/18237236/simplest_networking.zip)

Updated version of the project including the Networked red cube shown in EDIT 2 above. I'm gonna see if I can repro entirely without the Network Helper at all.

**Comment by agincel** [comment]
Okay, yeah. In another empty project, with *the only script* being an OnLoad to start a lobby (pulled out of Network Helper), I'm still seeing this behavior of Networked Objects destroyed on Host quit even if the settings suggest otherwise.

From left to right, these progressively paler red cubes are all Networked with their Orphaned modes set to
1. Destroy
2. Host
3. Random
4. Clear Owner

Upon the host leaving, all four cubes are destroyed. The fifth cube, which is just part of the network snapshot, is unaffected. So this has nothing to do with the fact that players own these objects, it's just nuking all objects upon the host leaving, as far as I can tell.

[even_simpler_networking.zip](https://github.com/user-attachments/files/18237307/even_simpler_networking.zip)
https://github.com/user-attachments/assets/9bdff3c1-9a00-4426-8248-d5786e41f926

**Comment by aylaylay** [comment]
Perfect repro

**Comment by agincel** [comment]
I only hope this behavior in "local hosting" mirrors what I was seeing in real hosted games. Harder to test those without a second machine set up - but considering in the past when I tested Marble Racer this way, quitting with the host process still let client processes roll around, I'm fairly confident it's a repro for the correct issue.

**Comment by aylaylay** [comment]
I don't think it's possible to test locally, the editor seems to just delete the scene without doing any of the orphan stuff

**Comment by SpiltBeanz** [comment]
This causes the ball to disappear in Fortwars as well when someone leaves the game. It just happens randomly when someone leaves. I'm not even doing host transfer so this still happens sometimes even when it's not the host leaving. 🤷‍♂

**Comment by aylaylay** [comment]
I have pushed a fix
